### ISSUE TITLE
Implement simple WL field level simulator (e.g. based on https://github.com/tlmakinen/kosmo-kompress )

### ISSUE THREAD

**Issue by EiffL** [astrodeepnet/sbi_experiments#18]

**Comment by tlmakinen** [comment]
thanks for tagging me ! I think I could easily incorporate this simulator into my compression framework for the IMNN (e.g. [](https://github.com/tlmakinen/kosmo-kompress)) to see how my SBI shakes out... would be great to have the k -> ell conversion eventually...

**Comment by EiffL** [comment]
Yep, I have played very quickly with training a simple compressor network, but training an IMNN would be nice. The main focus of this project is how to efficiently do the inference part, from "a" compressed statistics.

**Comment by EiffL** [comment]
After additional tinkering, it looks like it kind of works :-) learning the posterior distribution with combination of NLL and score loss. (notebook here )

So for instance, here is the true posterior distribution:
![image](https://user-images.githubusercontent.com/861591/148945001-143f5c27-4f81-4eb9-808f-4ebebadd1535.png)

Here is what you can estimate when using only 64 simulations and a NLL loss:
![image](https://user-images.githubusercontent.com/861591/148945062-0dbb962e-ca48-4c7a-88e6-c96fad898b21.png)

and here is what you get if you use 64 sims and a combination of NLL and score loss:
![image](https://user-images.githubusercontent.com/861591/148945135-8469a7bb-459f-4f24-9bba-82274f1e13ce.png)
Not yet perfect (it's only 64 sims) but looks like a huge improvement already.

So I think this example should be enough for our purposes at the timescale a ~a week.

**Comment by EiffL** [comment]
Did you manage to run the notebook @Justinezgh ? Did you run into any problems?

**Comment by Justinezgh** [comment]
Yes :) 
But I don't really understand why you use jax.lax.stop_gradient in the eval_model function when you compute the log p(theta | x, z)

**Comment by EiffL** [comment]
It's because you are evaluating the at the position of the point x that had just been sampled, but as if that was a "data point" coming from outside the model i.e. you don't want the gradients to propagate through x which should remain fixed.

**Comment by Justinezgh** [comment]
Since p(theta|x) = \int p(theta | x,z)p(z)dz, can I do this to get a 'real' approximation of the posterior : 

```python 
def get_jointlog_posterior(params, key):
    omega_c, s8 = params

    cond_model = condition(model, {'omega_c': 0.27, 'sigma_8': 0.77})
    cond_model = seed(cond_model, key)
    m_data = cond_model()
    
    cond_model = condition(model, {'omega_c': omega_c, 'sigma_8': s8, 'x': m_data})
    cond_model = seed(cond_model, key)
    model_trace = trace(cond_model).get_trace()
    jointlog = jnp.array([model_trace[k]['fn'].log_prob(model_trace[k]['value']) for k in model_trace.keys()]).sum()
    return jointlog


get_log_prob = jax.jit(jax.vmap(get_jointlog_posterior, in_axes=[0, None]))


p = jnp.stack(jnp.meshgrid(jnp.linspace(0.2,0.4,64),
                 jnp.linspace(0.7,0.9,64)), axis=-1)
ps = p.reshape([-1, 2])


true_posterior = jnp.array([np.stack([get_log_prob(params, jax.random.PRNGKey(k)) for params in p], axis=0) for k in range(50)]).mean(axis = 0)

```
![image](https://user-images.githubusercontent.com/72011736/152069989-8dabca09-4f75-4f73-9426-9baab05b028a.png)

Just to get an idea of what I'm supposed to have

**Comment by Justinezgh** [comment]
hmm I think this ^^^  is p(x, theta) so I have p(theta|x)*cte ?

**Comment by EiffL** [comment]
yeah it looks ok, but remember that it's an average over the probabilities, not the log probabilities ;-)

**Comment by Justinezgh** [comment]
So I trained 10 NFs for each nb of simulations (50,100,200,500,1000) and this is the results : 

![image](https://user-images.githubusercontent.com/72011736/152212410-c8752df3-2347-48c9-ba29-fe7b5904a076.png)

trained on simulations only : 

![ezgif-5-2d7e7b129d](https://user-images.githubusercontent.com/72011736/152239582-69cb538f-f942-493d-b0d3-cbbff9908915.gif)

trained on simulations and score : 

![ezgif-5-0b4635d982](https://user-images.githubusercontent.com/72011736/152239656-98a56c61-f6d5-47ef-93dc-58af9b9c4a59.gif)


maybe the pb can come from the weight on the score ? 

and I use the same random key to init all these NFs is that a pb ?

**Comment by Justinezgh** [comment]
I have the same behavior when I initialize the 10 NFs differently
![Unknown-12](https://user-images.githubusercontent.com/72011736/152324010-0830791c-7784-4401-ae10-9fb1c60f9078.png)

**Comment by Justinezgh** [comment]
Ok so 

1. I'm trying with another weight on the score (1500)

2. I'm doing the same thing with different metrics on the two moons to see if the pb come from the the metric and/or the stochasticity of the wl model 

3. Since we want to approximate the posterior sequentially at the end maybe 50 simulations are enough (and with this our posterior is much better than the one with simulations only). And as we take this posterior to become the new prior it will be easier for the nf to learn the posterior for a given x_0 since the parameter domain is smaller. Meaning that maybe for a set of parameters we will have several samples even if only 50 simulations are used. I don't know ^^

**Comment by EiffL** [comment]
These plots are really awesome, and yeah I think the first thing to try from here is to try adding starting from no score and doing a few runs with a increasing amount of score loss.

**Comment by Justinezgh** [comment]
nll + score/1500
![image](https://user-images.githubusercontent.com/72011736/152383954-ee0f16f8-cf78-42f1-9e1e-174268666e70.png)

**Comment by Justinezgh** [comment]
> These plots are really awesome, and yeah I think the first thing to try from here is to try adding starting from no score and doing a few runs with a increasing amount of score loss.

yes I'll try this

**Comment by EiffL** [comment]
Hummmmm it's still kinda surprising..... how do you compute the NLL in practice?

**Comment by Justinezgh** [comment]
negative log likelihood (weight = 0) : 

```python 
def loss_fn(params, batch, score, weight):
  y = compressor.apply(params_compressor, batch[1])
  log_prob, out = jax.vmap(jax.value_and_grad(lambda p, x: nvp.apply(params, p.reshape([-1,2]), x.reshape([-1,2]))))(batch[0], y) # Here we extract the grad of the model
  return -jnp.mean(log_prob) + jnp.mean(jnp.sum((out - score)**2, axis=1))*weight 

```

**Comment by Justinezgh** [comment]
negative log probability : 


``` python 
# Metric : neg log prob 
precision = 1000
oc = jnp.full(precision, 0.27)
s8 = jnp.full(precision, 0.77)
theta = jnp.stack([oc,s8], axis=1)

@jax.jit
def get_truth(theta, key):

  oc = theta[0]
  s8 = theta[1]

  cond_model = condition(seed(model, key), {'omega_c': oc, 'sigma_8': s8})
  model_trace = trace(cond_model).get_trace()
  return model_trace['x']['value']


# Negative log prob 
def neg_log_prob(params, params_compressor, theta, precision):

  master_seed = hk.PRNGSequence(1)
  x = jax.vmap(get_truth)(theta, jax.random.split(next(master_seed), precision))
  y = jax.vmap(lambda x: compressor.apply(params_compressor, x.reshape([-1,64,64])))(x)

  return - jnp.mean(jax.vmap((lambda p, x: nvp.apply(params, p.reshape([-1,2]), x.reshape([-1,2]))))(theta, y))
```

**Comment by Justinezgh** [comment]
for the two moons with 5 NFs for each nb of simu: 

![image](https://user-images.githubusercontent.com/72011736/152420585-6037e2b7-1557-4ab8-bee5-7c353128d501.png)
![image](https://user-images.githubusercontent.com/72011736/152422934-bb1e6c72-44e3-4f4c-8329-d9e24799ee9b.png)
![image](https://user-images.githubusercontent.com/72011736/152422961-14a9d1eb-b003-480a-ae09-2b19e7e84cbd.png)
![image](https://user-images.githubusercontent.com/72011736/152422993-0a4731d8-0ca8-4d28-a64a-aca86f56ecc3.png)

simulations only: 
![nllfinal](https://user-images.githubusercontent.com/72011736/152427360-ba7a3125-708b-4305-8920-81c8dbde19e0.gif)

simulations+score/1000:
![nllscorefinal](https://user-images.githubusercontent.com/72011736/152427376-6142a7e8-20f5-4eec-bdb7-8094c93e7aa3.gif)

**Comment by Justinezgh** [comment]
> These plots are really awesome, and yeah I think the first thing to try from here is to try adding starting from no score and doing a few runs with a increasing amount of score loss.

Starting with weight = 0 I increase it by 1e-4  every 1000 updates during 10000 updates
![image](https://user-images.githubusercontent.com/72011736/152427657-7b57c1fa-6244-42fd-a7db-a4bea9418b32.png)

**Comment by EiffL** [comment]
ah yeah no I mean, keeping weight constant during a given run, but doing several runs each one with a little bit more score loss.

But this plot is still interesting though... looks like it helps to start from a place where you train by NLL

**Comment by Justinezgh** [comment]
10 NFs for each nb of simulations (to get the uncertainty)
6 different losses corresponding to the 6 score weights: 0, 1e-3,5e-4,1e-4,5e-5,1e-5

![image](https://user-images.githubusercontent.com/72011736/152616877-2c1da453-cf87-4ce9-9630-1e84584e9569.png)

![image](https://user-images.githubusercontent.com/72011736/152617703-7adefdd2-fe51-47b9-ad50-f07c5c313665.png)

![image](https://user-images.githubusercontent.com/72011736/152617684-78dbee73-9cc6-4fac-a039-ce40fc28b527.png)

![image](https://user-images.githubusercontent.com/72011736/152617638-bea19312-f6fd-46af-8a80-5bf3c0c82baa.png)

![image](https://user-images.githubusercontent.com/72011736/152617782-a2e9052d-7108-49f6-a852-262dda83d7de.png)

![image](https://user-images.githubusercontent.com/72011736/152616947-fb39c360-e7ac-4734-a9a0-ae160b9d2b64.png)

**Comment by EiffL** [comment]
Fantastic Justine! Ok ok so, the story seems to be, adding a little bit of score loss helps, especially when we have a limited budget, but too much score loss is detrimental to the NLL.

Is the same thing true when learning just the two moons distribution is my question now

**Comment by EiffL** [comment]
oh sorry sorry, are these plots for the lensing simulator or the two moons?

**Comment by Justinezgh** [comment]
the lensing simulator

**Comment by Justinezgh** [comment]
for the two moons I have this : 
![](https://user-images.githubusercontent.com/72011736/152420585-6037e2b7-1557-4ab8-bee5-7c353128d501.png)

But I'll do the same thing with different weights to check

**Comment by Justinezgh** [comment]
did the same thing for the lensing simulator again just to check : 
![image](https://user-images.githubusercontent.com/72011736/152661707-ad3fb2e0-aae8-4b4f-9c6c-1e44dcad4973.png)
![image](https://user-images.githubusercontent.com/72011736/152661709-da82caca-797c-4131-bbdf-69d1402ab634.png)
![image](https://user-images.githubusercontent.com/72011736/152661712-82b93705-c335-4066-b626-c3b1f7e651e7.png)
![image](https://user-images.githubusercontent.com/72011736/152661716-9e178cac-cb80-4b58-b067-c90e292ebe6b.png)
![image](https://user-images.githubusercontent.com/72011736/152661720-43459d8a-7820-4c2c-87dc-820239f0c051.png)
![image](https://user-images.githubusercontent.com/72011736/152661727-66612a5f-20d7-42d6-bbf5-6e3e4fdd116f.png)

(I tried to do it with the two moons but I didn't take enough simulations to see the convergence at the end so tomorrow :) )

**Comment by Justinezgh** [comment]
same thing for the two moons (10NFs fpr each nb of simu)

![image](https://user-images.githubusercontent.com/72011736/152751874-5f67c296-5ecb-49dc-b53d-4c0e829774f1.png)
![image](https://user-images.githubusercontent.com/72011736/152751904-8517552f-8032-4cfa-a51a-5d61d5731464.png)
![image](https://user-images.githubusercontent.com/72011736/152751916-4bc09487-a732-411c-8c88-67a36d4e91eb.png)
![image](https://user-images.githubusercontent.com/72011736/152752034-c467eaba-d7d3-4be3-966a-aec1ed5b71b3.png)
![image](https://user-images.githubusercontent.com/72011736/152752055-a3f696a7-02e2-4b1d-8b74-250878d004ed.png)
![image](https://user-images.githubusercontent.com/72011736/152752072-38bce25c-bac7-4161-ab8e-f902aa2be58d.png)
